### PR TITLE
[Opt](orc)Optimize the merge io when orc reader read multiple tiny stripes.

### PR DESCRIPTION
**Pull request by hubgeter**
### What problem does this PR solve?

When reading orc files, we may encounter a scenario where the stripe byte size is very small but the number of stripes is very large. 

This pr introduces three session variables `orc_tiny_stripe_threshold_bytes`, `orc_once_max_read_bytes`, and `orc_max_merge_distance_bytes` to optimize io reading for the above scenarios. 

If a stripe byte size is less than `orc_tiny_stripe_threshold_bytes`, we will consider it as a tiny stripe. For multiple tiny stripes, we will perform IO merge reading according to the `orc_once_max_read_bytes` and `orc_max_merge_distance_bytes` parameters. Among them, `orc_once_max_read_bytes` indicates the maximum size of the merged IO. You should not set `orc_once_max_read_bytes` less than `orc_tiny_stripe_threshold_bytes`, although we will not force an error. When using tiny stripe reading optimization, since tiny stripes are not necessarily continuous, when the distance between two tiny stripes is greater than `orc_max_merge_distance_bytes`, we will not merge them into one IO.

If you don't want to use this optimization, you can `set orc_tiny_stripe_threshold_bytes = 0`.

Default parameters:
```mysql
orc_tiny_stripe_threshold_bytes = 8388608 (8M)
orc_once_max_read_bytes = 8388608 (8M)
orc_max_merge_distance_bytes = 1048576 (1M)
```

We also add relevant profiles for this purpose so that parameters can be adjusted to optimize reading.
`RangeCacheFileReader`:
1. `CacheRefreshCount`: how many IOs are merged
2. `ReadToCacheBytes`: how much data is actually read after merging
3. `ReadToCacheTime`: how long it takes to read data after merging
4. `RequestBytes`: how many bytes does the apache-orc library actually need to read the orc file
5. `RequestIO`: how many times the apache-orc library calls this read interface
6. `RequestTime`: how long it takes the apache-orc library to call this read interface

It should be noted that `RangeCacheFileReader` is a wrapper of the reader that actually reads data, such as the hdfs reader, so strictly speaking, `CacheRefreshCount` is not equal to how many IOs are initiated to hdfs, because each time the hdfs reader is requested, the hdfs reader may not be able to read all the data at once.

This pr also involves changes to the apache-orc third-party library: https://github.com/apache/doris-thirdparty/pull/244.
Reference implementation:  https://github.com/trinodb/trino/blob/master/lib/trino-orc/src/main/java/io/trino/orc/OrcDataSourceUtils.java#L36

#### Summary:
```mysql
set orc_tiny_stripe_threshold_bytes = xxx;
set orc_once_max_read_bytes = xxx;
set orc_max_merge_distance_bytes = xxx;

# xxx is the size in bytes
```

### Check List (For Committer)

- Test 

    - [x] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No colde files have been changed.
        - [ ] Other reason 

- Behavior changed:

    - [ ] No.
    - [x] Yes. 

- Does this need documentation?

    - [ ] No.
    - [x] Yes. 
doing...  

- Release note
   Introduces three session variables `orc_tiny_stripe_threshold_bytes`, `orc_once_max_read_bytes`, and `orc_max_merge_distance_bytes` to optimize io reading of scenarios where the orc stripe byte size is very small but the number of stripes is very large.

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 1]
```diff
@@ -1677,6 +1683,46 @@ public void setEnableLeftZigZag(boolean enableLeftZigZag) {
...
+                    + "我们认为该stripe为 tiny stripe。对于多个连续的tiny stripe我们会进行读取优化，即一次性读多个tiny stripe."
+                    + "如果你不想使用该优化，可以将该值设置为0。默认为 8M。",
+                    "In an orc file, if the byte size of a stripe is less than `orc_tiny_stripe_threshold`,"
+                            + "we consider the stripe to be a tiny stripe. For multiple consecutive tiny stripes,"
+                            + "we will perform read optimization, that is, read multiple tiny stripes at a time."
+                            + "If you do not want to use this optimization, you can set this value to 0."
+                            + "The default is 8M."},
+            needForward = true,
+            setter = "setOrcTinyStripeThreshold")
+    public long orcTinyStripeThreshold  = 8L * 1024L * 1024L;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:1698]:
```suggestion
    public long orcTinyStripeThresholdBytes  = 8L * 1024L * 1024L;
```

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 2]
```diff
@@ -1677,6 +1683,46 @@ public void setEnableLeftZigZag(boolean enableLeftZigZag) {
...
+            name = ORC_ONCE_MAX_READ_SIZE,
+            description = {"在使用tiny stripe读取优化的时候，会对多个tiny stripe合并成一次IO，"
+                    + "该参数用来控制每次IO请求的最大字节大小。你不应该将值设置的小于`orc_tiny_stripe_threshold`。默认为 8M。",
+                    "When using tiny stripe read optimization, multiple tiny stripes will be merged into one IO."
+                            + "This parameter is used to control the maximum byte size of each IO request."
+                            + "You should not set the value less than `orc_tiny_stripe_threshold`."
+                            + "The default is 8M."},
+            needForward = true,
+            setter = "setOrcOnceMaxReadSize")
+    public long orcOnceMaxReadSize = 8L * 1024L * 1024L;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:1711]:
```suggestion
    public long orcOnceMaxReadBytes = 8L * 1024L * 1024L;
```

File: fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java [hunk 3]
```diff
@@ -1677,6 +1683,46 @@ public void setEnableLeftZigZag(boolean enableLeftZigZag) {
...
+    @VariableMgr.VarAttr(
+            name = ORC_MAX_MERGE_DISTANCE,
+            description = {"在使用tiny stripe读取优化的时候，由于tiny stripe并不一定连续。"
+                    + "当两个tiny stripe之间距离大于该参数时，我们不会将其合并成一次IO。默认为 1M。",
+                    "When using tiny stripe read optimization, since tiny stripes are not necessarily continuous,"
+                            + "when the distance between two tiny stripes is greater than this parameter,"
+                            + "we will not merge them into one IO. The default value is 1M."},
+            needForward = true,
+            setter = "setOrcMaxMergeDistance")
+    public long orcMaxMergeDistance = 1024L * 1024L;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/qe/SessionVariable.java:1723]:
```suggestion
    public long orcMaxMergeDistanceBytes = 1024L * 1024L;
```

**Review by morningman — COMMENTED** [review]
Add test orc file

File: be/src/io/fs/buffered_reader.cpp [hunk 4]
```diff
@@ -869,5 +869,106 @@ Result<io::FileReaderSPtr> DelegateReader::create_file_reader(
                 return reader;
             });
 }
+
+Status LinearProbeRangeFinder::get_range_for(int64_t desiredOffset,
```
Comment (kaka11chen) [inline be/src/io/fs/buffered_reader.cpp:873]: desired_offset

File: be/src/io/fs/buffered_reader.h [hunk 5]
```diff
@@ -53,6 +53,121 @@ struct PrefetchRange {
...
+        }
+        result.push_back(last);
+        return result;
+    }
+};
+
+class RangeFinder {
+public:
+    virtual ~RangeFinder() = default;
+    virtual Status get_range_for(int64_t desiredOffset, io::PrefetchRange& result_range) = 0;
```
Comment (kaka11chen) [inline be/src/io/fs/buffered_reader.h:93]: `desired_offset `

File: be/src/io/fs/buffered_reader.h [hunk 6]
```diff
@@ -53,6 +53,121 @@ struct PrefetchRange {
...
+    virtual size_t get_max_range_size() const = 0;
+};
+
+class LinearProbeRangeFinder : public RangeFinder {
+public:
+    LinearProbeRangeFinder(std::vector<io::PrefetchRange>&& ranges) : _ranges(std::move(ranges)) {}
+
+    Status get_range_for(int64_t desiredOffset, io::PrefetchRange& result_range) override;
```
Comment (kaka11chen) [inline be/src/io/fs/buffered_reader.h:101]: `desired_offset `

File: be/src/io/fs/buffered_reader.h [hunk 7]
```diff
@@ -53,6 +53,121 @@ struct PrefetchRange {
...
+
+    PrefetchRange span(const PrefetchRange& other) const {
+        return {std::min(start_offset, other.end_offset), std::max(start_offset, other.end_offset)};
+    }
+    PrefetchRange seq_span(const PrefetchRange& other) const {
+        return {start_offset, other.end_offset};
+    }
+
+    //Ranges needs to be sorted.
+    static std::vector<PrefetchRange> mergeAdjacentSeqRanges(
```
Comment (kaka11chen) [inline be/src/io/fs/buffered_reader.h:65]: `merge_adjacent_seq_ranges()`

File: be/src/vec/exec/format/orc/vorc_reader.cpp [hunk 8]
```diff
@@ -853,6 +853,58 @@ Status OrcReader::set_fill_columns(
...
+                io::PrefetchRange::mergeAdjacentSeqRanges(
+                        tiny_stripe_ranges, orc_max_merge_distance, orc_once_max_read_size);
+
+        auto range_finder =
+                std::make_shared<io::LinearProbeRangeFinder>(std::move(prefetch_merge_ranges));
+
+        auto* orcInputStreamPtr = static_cast<ORCFileInputStream*>(_reader->getStream());
+        orcInputStreamPtr->set_all_tiny_stripes();
+        auto& orc_file_reader = orcInputStreamPtr->get_file_reader();
+        orc_file_reader->collect_profile_before_close();
```
[the remark this answers is not in the document]
Comment (hubgeter) [reply be/src/vec/exec/format/orc/vorc_reader.cpp:910]: Because before this, the reader needs to read the footer of the orc file

**Comment by hubgeter** [comment]
run buildall

File: be/src/vec/exec/format/orc/vorc_reader.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (hubgeter) [reply be/src/vec/exec/format/orc/vorc_reader.cpp]: I suddenly realized that I didn't need to call this thing

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Comment by hubgeter** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by kaka11chen — APPROVED** [review]
LGTM

**Pull request merged**